### PR TITLE
core : Float constructor allows to set binary or decimal precision

### LINKED ISSUE #12215: Float constructor should allow to set binary precision
I discussed this in some other issues, but I don't think I've opened an issue for it. 

The Float constructor only allows you to set the decimal precision. It then converts this to binary precision and saves that in `Float._prec`. 

The problem here is that there are fewer decimal precisions than binary precisions. For example:

```
In [9]: Float(1.0, 15)._prec
Out[9]: 53

In [10]: Float(1.0, 16)._prec
Out[10]: 56
```

So for instance, if you had a float that used 54 binary digits, it would be difficult to create a Float from it. I'm not even sure if it's possible, since in the current constructor, the `prec` argument overrides the fourth argument of a tuple input. 

One issue here is that everywhere else in the Sympy and mpmath code, `prec` refers to binary precision and `dps` refers to decimal precision. But currently, the second argument to `Float` is called `prec`, but refers to the decimal precision. So ideally, it should be 

```
Float(s, dps=15, prec=53)
```

where `Float(s, prec=54)` would override the default value of `dps=15`, and calling both (even consistently), like `Float(s, dps=15, prec=54)` would be an error.

Since the keyword argument name has to change, it would be a backwards compatibility break for anyone who calls Float like `Float(s, prec=20)`.  Ideally this would require a deprecation cycle, but it's not possible to detect if someone is calling `Float(s, prec=20)` meaning the old way (20 decimal digits) or the new way (20 binary digits). The good news is that `N` and `evalf` call this argument `n`, so there's no need to change things there.

### PR DESCRIPTION
**Pull request by ArifAhmed1995**
Either binary or decimal precision is supplied to the constructor of Float.
If neither is supplied the default precision of 15 decimal digits is used.
If both are supplied an error is thrown , even if decimal and binary precision
match.

Ref : #12215 
<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>
Fixes #12215

### CODE REVIEW

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -768,7 +768,11 @@ class Float(Number):
 
     is_Float = True
 
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None):
+        if dps and prec is not None:
+            raise ValueError('Both decimal and binary precision supplied.'
```
Comment (asmeurer) [inline sympy/core/numbers.py]: Need a space after the `.`

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -768,7 +768,11 @@ class Float(Number):
 
     is_Float = True
 
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None):
+        if dps and prec is not None:
```
Comment (asmeurer) [inline sympy/core/numbers.py]: Just `if dps and prec` would also work, and would read clearer.

Comment (asmeurer) [inline sympy/core/numbers.py]: And also in the code below, you can just check `prec` instead of `prec is None`, because it will evaluate to True if it is a nonzero integer (and 0 isn't a valid precision).

Comment (asmeurer) [inline sympy/core/numbers.py]: Oh I didn't know about the `prec=''` behavior. Never mind then.

**Comment by asmeurer** [comment]
This looks good. Can you add some tests for creating floats with precisions that aren't possible with dps?

**Comment by cbm755** [comment]
Cool!

A minor nitpick:
> If neither is supplied the default precision of 15 decimal digits is used.
This is not correct; the behaviour is more complicated than that.  More accurate statement is "... default precision of at least 15 decimal digits is used".

File: sympy/printing/tests/test_repr.py [hunk 3]
```diff
@@ -101,12 +101,13 @@ def test_Rational():
...
-       "Float('1.234567890123456789013', prec=19)")
-    sT(Float(
-        '0.60038617995049726', 15), "Float('0.60038617995049726', prec=15)")
+    sT(Float('1.23', dps=3), "Float('1.22998046875', prec=13)")
+    sT(Float('1.23456789', prec=33),
+            "Float('1.23456788994371891021728515625', prec=33)")
+    sT(Float('1.234567890123456789', prec=66),
+       "Float('1.23456789012345678901321800735590983322254032827913761138916015625', prec=66)")
+    sT(Float('0.60038617995049726', 15),
+       "Float('0.6003861799504972562857574303052388131618499755859375', prec=53)")
```
Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: This last test doesn't look right.  `prec=53` is roughly `dps = 15` right?  So why is it producing 53 decimal digits?  I think the last line should be
````
sT(Float('0.60038617995049726', 15), Float('0.60038617995049726', prec=53))
````
(or something very close to that).

**Review by cbm755 — CHANGES_REQUESTED** [review]
As I said, I like this idea, here are a few changes.

**Comment by asmeurer** [comment]
@cbm755 you are right. I didn't realize that `srepr` printed the keyword argument name. So this is a major compatibility break. Let's call the argument `_prec` and add a note that it should be changed in a future version.

**Comment by asmeurer** [comment]
(`_prec` seems the best name to me because that's also the name of the attribute)

**Comment by asmeurer** [comment]
OTOH `_prec` suggests it is a private API, which it isn't. So maybe `precision` is better. What do you guys think?

**Comment by cbm755** [comment]
`prec_`?   `precision` was the best I could come up with (see my comments at the original issue for more detail).

File: sympy/core/numbers.py [hunk 4]
```diff
@@ -784,11 +788,14 @@ def __new__(cls, num, prec=None):
         elif num is S.NegativeInfinity:
             num = '-inf'
         elif isinstance(num, mpmath.mpf):
-            if prec == None:
-                prec = num.context.dps
+            if prec is None:
+                if dps is None:
+                    prec = num.context.prec
+            else:
+                prec = num.context.prec
```
Comment (ArifAhmed1995) [inline sympy/core/numbers.py]: Ah, yes. I was just beginning to write some tests and I spotted this discrepancy.Thanks for pointing it out.

File: sympy/printing/tests/test_repr.py (hunk 3 above)
Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: Yes , the `_print_Float()` in `repr.py` uses the `to_str` function of mpmath. The `to_str` function uses decimal precision to yield the string representation of a mpmath float. 

I'll make the required changes.

File: sympy/core/numbers.py [hunk 5]
```diff
@@ -803,10 +821,11 @@ def __new__(cls, num, prec=None):
                     if num.is_Integer and isint:
                         dps = max(dps, len(str(num).lstrip('-')))
                     dps = max(15, dps)
-        elif prec == '':
+                    precision = mlib.libmpf.dps_to_prec(dps)
+        elif precision == '' and dps == None or precision == None and dps == '':
             if not isinstance(num, string_types):
                 raise ValueError('The null string can only be used when '
-                'the number to Float is passed as a string or an integer.')
+                'the number to Float is passed as a string or an integer. ')
```
Comment (cbm755) [inline sympy/core/numbers.py]: Remove spurious trailing space

File: sympy/core/numbers.py [hunk 6]
```diff
@@ -803,10 +821,11 @@ def __new__(cls, num, prec=None):
                     if num.is_Integer and isint:
                         dps = max(dps, len(str(num).lstrip('-')))
                     dps = max(15, dps)
-        elif prec == '':
+                    precision = mlib.libmpf.dps_to_prec(dps)
+        elif precision == '' and dps == None or precision == None and dps == '':
```
Comment (cbm755) [inline sympy/core/numbers.py]: 'is None' rather than =

File: sympy/printing/tests/test_repr.py [hunk 7]
```diff
@@ -101,12 +101,11 @@ def test_Rational():
 
 
 def test_Float():
-    sT(Float('1.23', prec=3), "Float('1.22998', prec=3)")
```
Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: I think these tests should stay almost "as is", just with dps=.  Then you can add your new tests as you've done here.  Logic here is nothing wrong with the old tests so leave then alone as Mich as possible .

**Review by cbm755 — CHANGES_REQUESTED** [review]
Looking good!

**Comment by cbm755** [comment]
Both of these should be explicitly disallowed
````
Float("1.23", dps="", precision=7)
Float("1.23", dps=5, precision="")
````

File: sympy/core/numbers.py [hunk 8]
```diff
@@ -768,7 +771,20 @@ class Float(Number):
 
     is_Float = True
 
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None, precision=None):
+        if dps and precision or dps and prec or prec and precision is not None:
```
Comment (cbm755) [inline sympy/core/numbers.py]: maybe you should move the `prec` deprecation stuff before this test so there is not `prec` in this if?

**Comment by cbm755** [comment]
We need some `raises(ValueError, lambda: Float("1.1", dps=3, precision=100)`, and also for the two cases I mentioned above.

File: sympy/core/numbers.py [hunk 9]
```diff
@@ -818,20 +837,21 @@ def __new__(cls, num, prec=None):
...
                         dps = max(dps, len(str(num).lstrip('-')))
+                        precision = mlib.libmpf.dps_to_prec(dps)
                     ok = True
             if ok is None:
                 raise ValueError('string-float not recognized: %s' % num)
-        else:
-            dps = prec
 
-        prec = mlib.libmpf.dps_to_prec(dps)
+        if precision is None:
```
Comment (cbm755) [inline sympy/core/numbers.py]: How about a comment here to help our future selves understand the logic?

````
# after the above, dps is definitely set and precision might be too.
assert dps is not None
if prec is None:
     ...
````

File: sympy/core/tests/test_numbers.py [hunk 10]
```diff
@@ -487,14 +488,30 @@ def teq(a):
...
+    assert Float(u'0.73908513321516064100000000') == \
+        Float('0.73908513321516064100000000')
+    assert Float(u'0.73908513321516064100000000', 28) == \
+        Float('0.73908513321516064100000000', 28)
+
+    # binary precision
+    # Decimal value 0.1 cannot be expressed precisely as a base 2 fraction
+    Float(S(1)/10, precision=53) == S(0.1)
+    Float(S(1)/10, precision=53) == S(0.1)
+    assert not Float(S(1)/10, precision=53) == Float(S(1)/10, precision=54)
```
Comment (ArifAhmed1995) [inline sympy/core/tests/test_numbers.py]: @asmeurer Is this the kind of test you talked about earlier ?

**Comment by ArifAhmed1995** [comment]
@cbm755 I've made the required changes.Can you please review again ?

**Review by cbm755 — CHANGES_REQUESTED** [review]

File: sympy/printing/tests/test_repr.py (hunk 7 above)
Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: In my view, it's better to always display in terms of binary precision to maintain the characteristics of the original Float. For example, we should print `Float('XXXX', precision = YYY)` if the Float was originally created with binary precision.

According to me(as per Sympy 1.0) , if I create : `Float('1.23', prec=3)` , I would expect the same to be printed out , because I explicitly mentioned `3` decimal digits.I don't see why it should be printed out as something different.But then again ,I may be incorrect in this.

I think that the value is stored correctly because both `a._mpf_` and `eval(srepr(a))._mpf_` are same.

Do you think it would be better if `srepr` prints with `dps=YY` when original Float is created with decimal precision and `precision=YY` for all other cases?

Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: @asmeurer Would like to hear your thoughts on this `srepr` issue.

Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: > it's better to always display in terms of binary precision 

Yes I agree.

My concern is that srepr printing has changed: before it was `Float('1.22998', prec=3)` now it is `Float('1.23', precision=13)`.  I agree that 13 is the right thing to print.  But why has displayed digits changed?  Its not necessarily bad; just not yet explained.

**Review by cbm755 — APPROVED** [review]

File: sympy/core/tests/test_numbers.py [hunk 11]
```diff
@@ -487,14 +488,30 @@ def teq(a):
...
+    assert Float(u'0.73908513321516064100000000') == \
+        Float('0.73908513321516064100000000')
+    assert Float(u'0.73908513321516064100000000', 28) == \
+        Float('0.73908513321516064100000000', 28)
+
+    # binary precision
+    # Decimal value 0.1 cannot be expressed precisely as a base 2 fraction
+    Float(S(1)/10, precision=53) == S(0.1)
+    Float(S(1)/10, precision=54) == S(0.1)
+    assert not Float(S(1)/10, precision=53) == Float(S(1)/10, precision=54)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: I think explicitly testing for this bullshit non-associate structural equality is a bad idea.  i.e., this is `a == b` and `c == b` but `a != c`.  This isn't your fault, but let's not build it into the tests.

Would this be a better test?
````

In [34]: a = Float(S(1)/10, dps=15)

In [35]: b = Float(S(1)/10, dps=16)

In [36]: p = Float(S(1)/10, precision=53)

In [37]: q = Float(S(1)/10, precision=54)

In [38]: a._mpf_ == p._mpf_
Out[38]: True

In [39]: a._mpf_ == q._mpf_
Out[39]: False

In [40]: b._mpf_ == q._mpf_
Out[40]: False
````
This shows that:
1.  `a` and `p` are the same
2.  But there are some bit precisions with no corresponding `dps`.  (`q` is "between" `a` and `b`).

File: sympy/printing/tests/test_repr.py (hunk 7 above)
Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: > Do you think it would be better if srepr prints with dps=YY when original Float is created with decimal precision and precision=YY for all other cases?

I don't think so; I like how you have it.

Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: It can be explained because an extra padding of 3 decimal digits is provided to make sure that no precision in lost between translations from decimal to binary(at least I think that's the case). Since now the translation from decimal to binary precision is taken care of by the `dps_to_prec` call in the `Float` constructor, there is no need for it.
```
def repr_dps(n):
    """Return the number of decimal digits required to represent
    a number with n-bit precision so that it can be uniquely
    reconstructed from the representation."""
    dps = prec_to_dps(n)
    if dps == 15:
        return 17
    return dps + 3
```

File: sympy/core/tests/test_numbers.py (hunk 11 above)
Comment (ArifAhmed1995) [inline sympy/core/tests/test_numbers.py]: Yes, this is a better test

File: sympy/printing/tests/test_repr.py (hunk 7 above)
Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: >  no precision in lost between translations from decimal to binary

Right, so that confirms why the extra 3 digits were there before.  Thanks for this.

> Since now the translation from decimal to binary precision is taken care

But I'm not sure the conclusion follows (that the extra digits are now unnecessary).  although I've yet to find a counterexample.

Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: `srepr` is used just to yield the string representation of the given Float, not for any further computation. I don't see why one would want to see extra decimal digits as compared to imput.

Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: You said it yourself:

> to make sure that no precision in lost between translations from decimal to binary

(This means between decimal and binary *representations* of the number, not between `dps` and `precision`.)

> not for any further computation.

No so, `eval(srepr(x))` should give exact same object whenever possible.

Comment (cbm755) [inline sympy/printing/tests/test_repr.py]: To elaborate: one might conclude that srepr of a float should prints its `_mpf`_`; but this is not so readable.  We want a similar guarantee of `eval(srepr(...))` being identity map.

Comment (ArifAhmed1995) [inline sympy/printing/tests/test_repr.py]: Ah yes , I now remember that the `sT()` function in `test_repr` does an evaluation on `srepr`.

File: sympy/printing/repr.py [hunk 12]
```diff
@@ -135,9 +135,8 @@ def _print_Fraction(self, expr):
         return 'Fraction(%s, %s)' % (self._print(expr.numerator), self._print(expr.denominator))
 
     def _print_Float(self, expr):
-        dps = prec_to_dps(expr._prec)
-        r = mlib.to_str(expr._mpf_, repr_dps(expr._prec))
-        return "%s('%s', prec=%i)" % (expr.__class__.__name__, r, dps)
+        r = mlib.to_str(expr._mpf_, prec_to_dps(expr._prec))
```
Comment (cbm755) [inline sympy/printing/repr.py]: Let's leave this as `repr_dps`.  You can always do another PR to change/discuss this, but it seems a shame to hold up this good `dps` change.

Its easier to get PRs that do one thing passed.  This current PR does two things: the dps/precision change and a (likely unrelated) srepr printing change.

File: sympy/core/tests/test_numbers.py [hunk 13]
```diff
@@ -1,4 +1,5 @@
 import decimal
+
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: extra blank line has snuck in

File: sympy/printing/repr.py (hunk 12 above)
Comment (ArifAhmed1995) [inline sympy/printing/repr.py]: Cool. I've done the changes.

Comment (ArifAhmed1995) [inline sympy/printing/repr.py]: @cbm755 Can this be merged now ? Another PR depends on the changes made here #11666

**Comment by cbm755** [comment]
> Can this be merged now ? Another PR(#11666) depends on the changes made here

I think so, you've addresses my concerns anyway and I think @asmeurer's comments as well.

I will merge in 24 hours or so.

While we wait, can you also draft some text [for the release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.1)?  (to note the API change.)

**Comment by ArifAhmed1995** [comment]
@cbm755 I've added the deprecation change to the release notes.

**Comment by asmeurer** [comment]
A deprecation issue was never opened for this. It's now https://github.com/sympy/sympy/issues/12820. Pull request adding it to the code here https://github.com/sympy/sympy/pull/12821.
